### PR TITLE
Coverage improvements

### PR DESCRIPTION
**Pull request by pjbakker**
Coverage improvements and some general bug fixes.

### CODE REVIEW

File: tests/suites/test_suite_md.function [hunk 1]
```diff
@@ -246,18 +258,28 @@ void md_hex_multi( char *text_md_name, char *hex_src_string,
...
     hexify( hash_str, output, mbedtls_md_get_size(md_info) );
+    TEST_ASSERT( strcmp( (char *) hash_str, hex_hash_string ) == 0 );
 
+    /* Test clone */
+    memset(hash_str, 0x00, 10000);
+    memset(output, 0x00, 100);
+
+    TEST_ASSERT ( 0 == mbedtls_md_update( &ctx_copy, src_str + halfway, src_len - halfway ) );
+    TEST_ASSERT ( 0 == mbedtls_md_finish( &ctx_copy, output ) );
+    hexify( hash_str, output, mbedtls_md_get_size(md_info) );
```
Comment (yanesca) [inline tests/suites/test_suite_md.function]: minor style issue: mbedtls_md_get_size(md_info) -> mbedtls_md_get_size( md_info )

File: tests/suites/test_suite_md.function [hunk 2]
```diff
@@ -207,15 +208,25 @@ void md_text_multi( char *text_md_name, char *text_src_string,
...
     hexify( hash_str, output, mbedtls_md_get_size(md_info) );
+    TEST_ASSERT( strcmp( (char *) hash_str, hex_hash_string ) == 0 );
 
+    /* Test clone */
+    memset(hash_str, 0x00, 1000);
+    memset(output, 0x00, 100);
+
+    TEST_ASSERT ( 0 == mbedtls_md_update( &ctx_copy, src_str + halfway, len - halfway ) );
+    TEST_ASSERT ( 0 == mbedtls_md_finish( &ctx_copy, output ) );
+    hexify( hash_str, output, mbedtls_md_get_size(md_info) );
```
Comment (yanesca) [inline tests/suites/test_suite_md.function]: minor style issue: mbedtls_md_get_size(md_info) -> mbedtls_md_get_size( md_info )

File: tests/suites/test_suite_md.function [hunk 3]
```diff
@@ -207,15 +208,25 @@ void md_text_multi( char *text_md_name, char *text_src_string,
...
 
     TEST_ASSERT ( 0 == mbedtls_md_starts( &ctx ) );
     TEST_ASSERT ( ctx.md_ctx != NULL );
     TEST_ASSERT ( 0 == mbedtls_md_update( &ctx, src_str, halfway ) );
+    TEST_ASSERT ( 0 == mbedtls_md_clone( &ctx_copy, &ctx ) );
+
     TEST_ASSERT ( 0 == mbedtls_md_update( &ctx, src_str + halfway, len - halfway ) );
     TEST_ASSERT ( 0 == mbedtls_md_finish( &ctx, output ) );
-
     hexify( hash_str, output, mbedtls_md_get_size(md_info) );
```
Comment (yanesca) [inline tests/suites/test_suite_md.function]: minor style issue: mbedtls_md_get_size(md_info) -> mbedtls_md_get_size( md_info )

File: tests/suites/test_suite_md.function [hunk 4]
```diff
@@ -190,9 +190,10 @@ void md_text_multi( char *text_md_name, char *text_src_string,
...
 
     const mbedtls_md_info_t *md_info = NULL;
-    mbedtls_md_context_t ctx;
+    mbedtls_md_context_t ctx, ctx_copy;
 
     mbedtls_md_init( &ctx );
+    mbedtls_md_init( &ctx_copy );
 
     memset(md_name, 0x00, 100);
     memset(src_str, 0x00, 1000);
```
Comment (yanesca) [inline tests/suites/test_suite_md.function]: Missing spaces around the parameters in both lines.

File: tests/suites/test_suite_md.function [hunk 5]
```diff
@@ -233,9 +244,10 @@ void md_hex_multi( char *text_md_name, char *hex_src_string,
...
     int src_len, halfway;
     const mbedtls_md_info_t *md_info = NULL;
-    mbedtls_md_context_t ctx;
+    mbedtls_md_context_t ctx, ctx_copy;
 
     mbedtls_md_init( &ctx );
+    mbedtls_md_init( &ctx_copy );
 
     memset(md_name, 0x00, 100);
     memset(src_str, 0x00, 10000);
```
Comment (yanesca) [inline tests/suites/test_suite_md.function]: Missing spaces around the parameters in both lines.

File: tests/suites/test_suite_md.function [hunk 6]
```diff
@@ -246,18 +258,28 @@ void md_hex_multi( char *text_md_name, char *hex_src_string,
...
+
     TEST_ASSERT ( 0 == mbedtls_md_update( &ctx, src_str + halfway, src_len - halfway) );
     TEST_ASSERT ( 0 == mbedtls_md_finish( &ctx, output ) );
-
     hexify( hash_str, output, mbedtls_md_get_size(md_info) );
+    TEST_ASSERT( strcmp( (char *) hash_str, hex_hash_string ) == 0 );
 
+    /* Test clone */
+    memset(hash_str, 0x00, 10000);
+    memset(output, 0x00, 100);
```
Comment (yanesca) [inline tests/suites/test_suite_md.function]: Missing spaces around the parameters in both lines.

File: tests/suites/test_suite_md.function [hunk 7]
```diff
@@ -207,15 +208,25 @@ void md_text_multi( char *text_md_name, char *text_src_string,
...
+
     TEST_ASSERT ( 0 == mbedtls_md_update( &ctx, src_str + halfway, len - halfway ) );
     TEST_ASSERT ( 0 == mbedtls_md_finish( &ctx, output ) );
-
     hexify( hash_str, output, mbedtls_md_get_size(md_info) );
+    TEST_ASSERT( strcmp( (char *) hash_str, hex_hash_string ) == 0 );
 
+    /* Test clone */
+    memset(hash_str, 0x00, 1000);
+    memset(output, 0x00, 100);
```
Comment (yanesca) [inline tests/suites/test_suite_md.function]: Missing spaces around the parameters in both lines.

File: tests/suites/test_suite_md.function [hunk 8]
```diff
@@ -246,18 +258,28 @@ void md_hex_multi( char *text_md_name, char *hex_src_string,
...
 
     TEST_ASSERT ( 0 == mbedtls_md_starts( &ctx ) );
     TEST_ASSERT ( ctx.md_ctx != NULL );
     TEST_ASSERT ( 0 == mbedtls_md_update( &ctx, src_str, halfway ) );
+    TEST_ASSERT ( 0 == mbedtls_md_clone( &ctx_copy, &ctx ) );
+
     TEST_ASSERT ( 0 == mbedtls_md_update( &ctx, src_str + halfway, src_len - halfway) );
     TEST_ASSERT ( 0 == mbedtls_md_finish( &ctx, output ) );
-
     hexify( hash_str, output, mbedtls_md_get_size(md_info) );
```
Comment (yanesca) [inline tests/suites/test_suite_md.function]: minor style issue: mbedtls_md_get_size(md_info) -> mbedtls_md_get_size( md_info )

File: tests/suites/test_suite_asn1write.data [hunk 9]
```diff
@@ -67,14 +67,26 @@ mbedtls_asn1_write_len:255:"81FF":2:2
 ASN.1 Write / Read Length #5 (Len = 255, buffer too small)
 mbedtls_asn1_write_len:255:"81FF":1:MBEDTLS_ERR_ASN1_BUF_TOO_SMALL
 
-ASN.1 Write / Read Length #6 (Len = 256, long form)
-mbedtls_asn1_write_len:256:"820100":3:3
+ASN.1 Write / Read Length #6 (Len = 258, byte order)
```
Comment (yanesca) [inline tests/suites/test_suite_asn1write.data]: What does byte order mean here?

**Comment by yanesca** [comment]
+1

File: tests/suites/test_suite_asn1write.data (hunk 9 above)
Comment (pjbakker) [inline tests/suites/test_suite_asn1write.data]: With 0xFFFF you cannot see if the bytes were place in the correct order into the ASN1 representation.  With 0x0102 you can.

-------- Original message --------

File: tests/suites/test_suite_md.function (hunk 4 above)
Comment (pjbakker) [inline tests/suites/test_suite_md.function]: In the lines already there.. You are right of course, but I wasn't necessarily doing a run of fixing things in the code that were already present.. But I'll do it now anyway..

File: tests/suites/test_suite_cipher.aes.data [hunk 10]
```diff
@@ -1097,3 +1097,6 @@ test_vec_ecb:MBEDTLS_CIPHER_AES_256_ECB:MBEDTLS_DECRYPT:"00000000000000000000000
 AES-256-ECB Decrypt NIST KAT #12
 depends_on:MBEDTLS_AES_C
 test_vec_ecb:MBEDTLS_CIPHER_AES_256_ECB:MBEDTLS_DECRYPT:"0000000000000000000000000000000000000000000000000000000000000000":"9b80eefb7ebe2d2b16247aa0efc72f5d":"e0000000000000000000000000000000":0
+
+Cipher Corner Case behaviours
```
Comment (sbutcher-arm) [inline tests/suites/test_suite_cipher.aes.data]: Looks like this special case should be dependent on MBEDTLS_AES_C.

**Comment by sbutcher-arm** [comment]
Reviewed, all issues addressed, so ready for merge.

File: tests/suites/test_suite_cipher.aes.data (hunk 10 above)
Comment (pjbakker) [inline tests/suites/test_suite_cipher.aes.data]: Please see line 95 in the patch to test_suit_cipher.function where the dependency is stated..
